src/fciqmc.jl

```diff
@@ -481,53 +520,151 @@ function fciqmc_step!(Ĥ, dv, shift, dτ, pnorm, ws::NTuple{NT,W};
 
-function Rimu.fciqmc_step!(Ĥ, dv::MPIData{D,S}, shift, dτ, pnorm, w::D;
-                           m_strat::MemoryStrategy = NoMemory()) where {D,S}
-    # MPI version, single thread
+import SplittablesBase.halve, SplittablesBase.amount
+import Base.Threads.@spawn, Base.Threads.nthreads, Base.Threads.threadid
+
+# This version seems to be faster but have slightly more allocations
+function fciqmc_step!(Ĥ, dv, shift, dτ, pnorm, ws::NTuple{NT,W}, f::Float64;
+      m_strat::MemoryStrategy = NoMemory()
+    ) where {NT,W}
+    # multithreaded version; should also work with MPI
+    @assert NT == nthreads() "`nthreads()` not matching dimension of `ws`"
     v = localpart(dv)
-    @assert w ≢ v "`w` and `v` must not be the same object"
-    empty!(w)
-    stats = zeros(Int, 5) # pre-allocate array for stats
-    for (add, num) in pairs(v)
-        res = Rimu.fciqmc_col!(w, Ĥ, add, num, shift, dτ)
-        stats .+= res # just add all stats together
+    statss = allocate_statss(v, NT)
+
+    batchsize = max(100.0, min(amount(pairs(v))/NT, sqrt(amount(pairs(v)))*10))
+
+    # define recursive dispatch function that loops two halves in parallel
+    function loop_configs!(ps) # recursively spawn threads
+        if amount(ps) > batchsize
+            two_halves = halve(ps) #
+            fh = @spawn loop_configs!(two_halves[1]) # runs in parallel
+            loop_configs!(two_halves[2])           # with second half
+            wait(fh)                             # wait for fist half to finish
+        else # run serial
+            # id = threadid() # specialise to which thread we are running on here
+            # serial_loop_configs!(ps, ws[id], statss[id], trng())
+            for (add, num) in ps
+                ss = fciqmc_col!(ws[threadid()], Ĥ, add, num, shift, dτ)
+                # @show threadid(), ss
+                statss[threadid()] .+= ss
+            end
+        end
+        return nothing
     end
-    r = applyMemoryNoise!(w, v, shift, dτ, pnorm, m_strat) # memory noise
-    # thresholdProject!(w, v, shift, dτ, m_strat) # apply walker threshold if applicable
-    sort_into_targets!(dv, w)
-    MPI.Allreduce!(stats, +, dv.comm) # add stats of all ranks
-    return dv, w, stats, r
-    # returns the structure with the correctly distributed end
-    # result `dv` and cumulative `stats` as an array on all ranks
-    # stats == (spawns, deaths, clones, antiparticles, annihilations)
+
+    # function serial_loop_configs!(ps, w, rng, stats)
+    #     @inbounds for (add, num) in ps
+    #         ss = fciqmc_col!(w, Ĥ, add, num, shift, dτ, rng)
+    #         # @show threadid(), ss
+    #         stats .+= ss
+    #     end
+    #     return nothing
+    # end
+
+    zero!.(ws) # clear working memory
+    loop_configs!(pairs(v))
+
+    r = applyMemoryNoise!(ws, v, shift, dτ, pnorm, m_strat) # memory noise
+    return (sort_into_targets!(dv, ws, statss)... , r) # MPI syncronizing
+    #
+    # return statss
+end
+
+using ThreadsX
```

Comment:
This was already imported in `src/Rimu.jl`

Reply:
fixed